### ISSUE TITLE
PackageKit restarts computer twice to apply offline updates

### ISSUE THREAD

**Issue by raploz** [PackageKit/PackageKit#510]
This is not technically a bug but a design decision that degrades the UX of users trying to update their systems. The way PackageKit applies updates could be improved to make updating less frustrating to users.

On Windows, when the user is prompted to restart their system to update, part of the updates are applied on shutdown and the final steps are applied when booting up the next time. When the system finishes installing updates it goes straight into the desktop. While this can't be considered a seamless experience, it's faster than what PackageKit currently does.

On Linux with PackageKit, when the user is prompted to restart, the system restarts and updates are applied on the next start-up. After it finishes, it restarts once again and only then it boots up normally, thus adding additional time until the computer is ready to use. It's important to remember an additional restart is problematic because now the user has to wait twice for the BIOS screen and GRUB, so this adds up quickly.

Another problem is, some distros choose offline updates by default because they usually update often (e.g Fedora or KDE Neon) and online updates causes bugs. Having to update every day and waiting for the computer to boot twice can quickly become a problem. Since one of the downsides of online/in-system updates are bugs, those distributions won't change their defaults. There's no alternative but to improve offline updates to make them as quick as possible. It's not feasible to expect those distros to change their defaults due to those reasons.

I believe PackageKit should work in one of three ways:

1. Keep applying offline updates on the next reboot, but instead of rebooting again, finish the process and boot normally.
2. Apply updates while shutting down instead of when turning on. The next time the computer starts up, do the final steps and load straight into the desktop without rebooting again.
or
3. Apply updates in a temporary location in the background while the computer is still in use (so it won't overwrite currently running programs). If the user restarts their computer while operations are still pending, finish the process while turning off and then restart. On the next reboot, updates should overwrite the old files, and the computer boots straight into the desktop without needing an additional restart.

Even if none of these options are feasible, I believe this should still be discussed, as updates are one of the main pain points of Windows users, and making the process take even longer on Linux increases downtime and is detrimental to user experience.

Thank you for your attention.

**Comment by www-cat** [comment]
could `kexec` or `systemd isolate` be used to reboot into pk without power cycling?

i tried the former on debian but it rebooted into desktop.

**Comment by Lurux** [comment]
@hughsie sorry to necropost here, but I'd like to provide more input on this.

___

Two problems that were not mentioned here are that GRUB may boot into another OS in a multi-boot configuration, and the LUKS password will be asked before updating encrypted systems.
This may be a problem if the user clicks "update and shut down" and goes away from the computer expecting it to do its thing.

As far as boot "pristineness" goes, wouldn't it be possible completely restart systemd, without actually rebooting the computer ?
In other words, stop all services (like with `shutdown.target`), then restart systemd and boot it to `system-update.target` (maybe using `daemon-reload`) ?
This should address most problems on the userspace side, AFAIK.

If there are any concerns on the kernel side of things, I know kexec can be used to boot another kernel from a running system; I assume it could also be used to reload the current kernel, which should rerun any kernel-side initialization, as I understand it ?
I think this would address your concerns while skipping firmware and GRUB, but maybe it's not worth the effort at that point ?

___

**TLDR**

Could `kexec` or some kind of `systemctl daemon-restart` be used to reboot into offline updates without power cycling ?

___

Hoping this message was useful and didn't come as an annoyance,
Cheers !

**Comment by Lurux** [comment]
I suspected that, hence the "maybe not worth the effort" comment - though, it would still fix the GRUB problem, and significantly reduce the time before the LUKS prompt, somewhat reducing the friction of "update and shut down".

All of that is assuming a complete systemd restart wouldn't be sufficient.

**Comment by Pointedstick** [comment]
>> Apply updates while shutting down instead of when turning on
>
> Also not possible as you need a pristine boot to apply updates safely, rather than a system that has been futzed with.

Could we have PackageKit do the following in response to an "Install updates and shut down" command from the user:
1. Reboot into the clean environment
2. Run the update
3. Shut down instead of rebooting

**Comment by ximion** [comment]
Yes we could - but I would much rather change the implementation completely, at least on systemd-systems:
* Soft-reboot the current running OS with systemd (so we keep all encrypted devices unlocked, but do set the system into a state that's as predictable as possible)
* Install updates
* Hard reboot or poweroff

**Comment by ximion** [comment]
I do not think this feature is supported on non-systemd systems at all, but I haven't looked into it recently, maybe support was added.

**Comment by hughsie** [comment]
> Soft-reboot the current running OS with systemd

I think that's harder than you think; the idea of a fresh reboot was that you've not got various bad things happening to your hardware -- e.g. kernel drivers that have oops, random zombie processes etc.

**Comment by ximion** [comment]
> I think that's harder than you think; the idea of a fresh reboot was that you've not got various bad things happening to your hardware -- e.g. kernel drivers that have oops, random zombie processes etc.

I actually wanted to ask for your feedback before making any such change anyway (and also hope someone sponsors that kind of work ^^).
There has been an incredible number of complaints about the upgrade experience, and I would like to improve it, but there's just no getting around moving the system into some kind of safe state before running the update (lots of tickets and broken systems have taught me that - users can be pretty reckless too, besides something in the system going wrong). 

Is the driver bugs a common thing? I was thinking of potentially augmenting systemd's soft reboot with a "soft reboot into a minimal-amount-of-daemons-running" stage, to perform the update in. The kernel oops IMHO would be just as likely after a reboot, if it's a concern at that stage.
The zombie processes and all other clutter should (hopefully, if there are no bugs) all have been taken care of by systemd. I doubt that without systemd anything comparable would even be possible.

The advantage is that the user isn't annoyed on reboot when they want to use their system with a new update, people don't need to babysit their update when they use disk encryption (important for Purism/PureOS!) and the experience should be overall a bit smoother.
Of course, the kernel will not be reset when the update is run, and remain in its previous state with that scheme.

**Comment by maverick74** [comment]
@ximion do you think [systemd's v254 soft reboot](https://media.ccc.de/v/all-systems-go-2023-184-soft-reboot-atomically-replace-rootfs-and-reboot-userspace-without-kernel-restart) helps this in anyway?

Would be great to have this moving forward!

**Comment by maverick74** [comment]
> i tried `pkcon offline-trigger` and `systemctl soft-reboot`, and it updated and rebooted only once! 

But what about Update & ShutDown?
